drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/main/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTablePopulater.java

```diff
@@ -116,70 +118,131 @@ private void processIndirectSourceBuilders(final int maxRowCount) {
+        final List<BRLVariableColumn> variableColumns = new ArrayList<>();
+        final List<GuidedDecisionTableLHSBuilder> lhsBuilders = new ArrayList<>();
+        final List<GuidedDecisionTableRHSBuilder> rhsBuilders = new ArrayList<>();
         for (GuidedDecisionTableSourceBuilder sb : sourceBuilders) {
             if (sb instanceof GuidedDecisionTableSourceBuilderIndirect) {
                 for (BRLVariableColumn variableColumn : ((GuidedDecisionTableSourceBuilderIndirect) sb).getVariableColumns()) {
                     variableColumns.add(variableColumn);
                 }
             }
-        }
-
-        //Convert the DRL to a RuleModel from which we construct BRLFragment columns
-        final StringBuilder rule = new StringBuilder();
-        if (!(dmo.getPackageName() == null || dmo.getPackageName().isEmpty())) {
-            rule.append("package ").append(dmo.getPackageName()).append("\n");
-        }
-        rule.append("rule 'temp' \n").append("when \n");
-        for (GuidedDecisionTableSourceBuilder sb : sourceBuilders) {
             if (sb instanceof GuidedDecisionTableLHSBuilder) {
-                rule.append(sb.getResult());
+                lhsBuilders.add((GuidedDecisionTableLHSBuilder) sb);
             }
-        }
-        rule.append("\nthen \n");
-        for (GuidedDecisionTableSourceBuilder sb : sourceBuilders) {
             if (sb instanceof GuidedDecisionTableRHSBuilder) {
-                rule.append(sb.getResult());
+                rhsBuilders.add((GuidedDecisionTableRHSBuilder) sb);
             }
         }
-        rule.append("end");
-        final RuleModel rm = RuleModelDRLPersistenceImpl.getInstance().unmarshal(rule.toString(),
-                                                                                 Collections.emptyList(),
-                                                                                 dmo);
-        if (rm.lhs != null) {
-            for (IPattern pattern : rm.lhs) {
+
+        //-- PROCESS LHS PATTERNS
+        // For each GuidedDecisionTableLHSBuilder gather IPatterns generated
+        // from each accumulative iteration and convert to RuleModel
+        int previousPatternCount = 0;
+        for (int i = 0; i < lhsBuilders.size(); i++) {
+            //Construct DRL for this and previous builders
+            final StringBuilder rule = new StringBuilder();
+            if (!(dmo.getPackageName() == null || dmo.getPackageName().isEmpty())) {
+                rule.append("package ").append(dmo.getPackageName()).append("\n");
+            }
+            rule.append("rule 'temp' \n").append("when \n");
+            for (int inner = 0; inner <= i; inner++) {
+                final GuidedDecisionTableLHSBuilder lhsBuilder = lhsBuilders.get(inner);
+                rule.append(lhsBuilder.getResult());
+            }
+            rule.append("\nthen \n");
+            rule.append("end");
```

Comment:
General question for whole class. Sometimes we add ' ' before the new line character, sometimes not. Also do not we want to use `System.lineSeparator()` instead of '\n' ? Just asking I do not know.

Reply:
Where and why we add line separators is a bit of black magic in this (old) class.

I'd rather not analyse as part of this JIRA/PR as it runs the risk of breaking things.

The marshaller [splits](https://github.com/kiegroup/drools/blob/master/drools-workbench-models/drools-workbench-models-commons/src/main/java/org/drools/workbench/models/commons/backend/rule/RuleModelDRLPersistenceImpl.java#L1972) on `\n` so I'd rather leave "as is".

If you feel we should address both issues could you please make a new JIRA.